enhancements/insights/prometheus-alerts-for-insights-recommendations.md

```diff
@@ -0,0 +1,268 @@
+  - Reset the Insights recommendation metrics to avoid showing alerts about
+    recommendations that are no longer hitting.
+- Unable to fetch a report from Smart Proxy (endpoint unreachable, auth error, etc.)
+  - Log error, do not touch the Insights recommendation metrics.
+- The fetched report cannot be decoded into the expected structure, or the data
+  are otherwise invalid
+  - Log error, do not touch the Insights recommendation metrics.
+
+If there is anything wrong with the Smart Proxy or the reports returned by it,
+the metrics will not be modified until the validity of the last report expires.
+If IO is enabled, but something is wrong with the gathering or uploading, the
+metrics will not be directly affected for as long as the last report is valid.
+If IO has been disabled, or crashing for more then 10 minutes, the alerts will
+be disabled to avoid presenting the user with outdated information.
+
+### Risks and Mitigations
+
+- Disabling and re-enabling an alert (e.g, due to IO or Smart Proxy outage)
+  can trigger user-defined actions, even though there has not been any actual
+  change to the status of the recommendation. This could be very annoying if
+  it were to happen often or repeatedly over a short period of time.
+- The URL to detailed information about an active Insights recommendation will
+  be hard-coded into the alert template or the Insights Operator. If the
+  endpoint moves elsewhere, the URL will be broken and an updated version of the
+  alert template will have to be backported. The process can take several
+  months, and there is no guarantee that the user will perform the update.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Which operator conditions should fall under the unhealthy Insights Operator
+   state?
+1. How long should IO wait before considering a report to be no longer valid
+   (outdated)?
+1. If uploads keep failing, how and when will the alerts be disabled? (because
+   the current design assumes that the metrics are only modified after receiving
+   a report after a successful upload)
+
+### Test Plan
+
+Due to its nature, this feature will probably be best tested using e2e tests
+that involve other components. The Prometheus stack will be necessary to
+evaluate the metrics and alerts.
+
+The e2e tests could work as follows:
+- break some property of the cluster,
+- have the Insights Operator gather data about it,
+- upload the Insights archive,
+- allow the CCX data pipeline to process the archive,
+- check the resulting metrics/alerts.
+
+### Graduation Criteria
+
+N/A
+
+This feature does not introduce a new API.
+
+There will first be a PoC to verify viability of the feature, and then the next
+step will be the full implementation ready for release.
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+Downgrading to an older version will remove the alert template definition, if
```

Comment:
this doesn't happen automatically; you'd need to teach the alert-resource-creator in the older release to know that it should remove the alerts.  If it's the CVO pushing those rules on your behalf, that means manifests with the [`release.openshift.io/delete` annotation][1].  On the other hand, [rollbacks are not supported][2], and with the older Insights operator no longer serving the backing metrics, having the stale alert rule lying around doesn't seem like that big a deal.

[1]: https://github.com/openshift/enhancements/blob/20183fad84f682c83e60386593c0eca717ee5bc9/enhancements/update/object-removal-manifest-annotation.md#proposal
[2]: https://github.com/openshift/openshift-docs/blame/d4762f0f626a4dddb9d7330e63a3bb6cb73f5bb5/modules/update-upgrading-cli.adoc#L160-L162

Reply:
Okay, I guess this is a non-issue either way, then. IO is managed by CVO, and the alert definitions would be shipped with it, so presumably also managed by CVO. Thanks for pointing this out, and I believe it should be fine to add that annotation to the alert template when we write it.